src/autopas/LogicHandler.h

```diff
@@ -1772,42 +1771,93 @@ void LogicHandler<Particle_T>::remainderHelper3bBufferBufferContainerAoS(
   const double cutoff = container.getCutoff();
+  const double cutoffSquared = cutoff * cutoff;
+
+  // Step 1: Create a sorted cell view on all buffer particles
+  const auto domainDiagonal = container.getBoxMax() - container.getBoxMin();
+  ReferenceParticleCell<Particle_T> cellView;
+  cellView.reserve(numBufferParticles);
+  for (auto pRef : bufferParticles) {
+    cellView.addParticleReference(pRef);
+  }
+  SortedCellView<ReferenceParticleCell<Particle_T>> sortedCellView(cellView, domainDiagonal);
 
-  for (auto i = 0; i < bufferParticles.size(); ++i) {
-    Particle_T &p1 = *bufferParticles[i];
-    const auto pos = p1.getR();
-    const auto boxmin = pos - cutoff;
-    const auto boxmax = pos + cutoff;
+  // Step 2: Create a neighbor list of pairs for each of the buffer particles.
+  // This step accesses all particles in read-only mode and can be fully parallelized.
+  std::vector<std::vector<std::pair<Particle_T *, Particle_T *>>> bufferParticleNeighborLists(numBufferParticles);
+
+  AUTOPAS_OPENMP(parallel for schedule(static))
+  for (size_t i = 0; i < numBufferParticles; i++) {
+    auto &[pProjection, p1Ptr] = sortedCellView._particles[i];
+    const auto pos1 = p1Ptr->getR();
+    const auto boxMin = pos1 - cutoff;
+    const auto boxMax = pos1 + cutoff;
 
     auto p2Iter = container.getRegionIterator(
-        boxmin, boxmax, IteratorBehavior::ownedOrHalo | IteratorBehavior::forceSequential, nullptr);
+        boxMin, boxMax, IteratorBehavior::ownedOrHalo | IteratorBehavior::forceSequential, nullptr);
     for (; p2Iter.isValid(); ++p2Iter) {
-      Particle_T &p2 = *p2Iter;
-
+      const auto pos2 = p2Iter->getR();
+      const auto p1p2Dist = pos2 - pos1;
+      if (utils::ArrayMath::dot(p1p2Dist, p1p2Dist) > cutoffSquared) {
+        continue;
+      }
       auto p3Iter = p2Iter;
       ++p3Iter;
-
       for (; p3Iter.isValid(); ++p3Iter) {
-        Particle_T &p3 = *p3Iter;
+        const auto pos3 = p3Iter->getR();
+        const auto p2p3Dist = pos3 - pos2;
+        if (utils::ArrayMath::dot(p2p3Dist, p2p3Dist) > cutoffSquared) {
+          continue;
+        }
+        const auto p1p3Dist = pos3 - pos1;
+        if (utils::ArrayMath::dot(p1p3Dist, p1p3Dist) > cutoffSquared) {
+          continue;
+        }
+        // All distances are below the cutoff, so append to the neighbor list.
+        bufferParticleNeighborLists[i].emplace_back(&(*p2Iter), &(*p3Iter));
+      }
+    }
+  }
 
-        if constexpr (newton3) {
-          f->AoSFunctor(p1, p2, p3, true);
-        } else {
-          if (i < numOwnedBufferParticles) {
-            f->AoSFunctor(p1, p2, p3, false);
-          }
-          if (p2.isOwned()) {
-            f->AoSFunctor(p2, p1, p3, false);
-          }
-          if (p3.isOwned()) {
-            f->AoSFunctor(p3, p1, p2, false);
+  // Step 3: Bin buffer particles into slices along the domain diagonal of size cutoff
+  std::vector<std::vector<std::pair<size_t, Particle_T *>>> binnedBufferParticles;
```

Comment:
Again, explain what each part of the structure is for. (And potentially keep in memory)

Reply:
The logic for this step needs explaining in a brief comment.